File: 833.字符串中的查找与替换.cpp
/*
 * @lc app=leetcode.cn id=833 lang=cpp
 *
 * [833] 字符串中的查找与替换
 */

// @lc code=start
#include<vector>
#include<string>
#include<tuple>
#include<algorithm>
#include<iostream>
using namespace std;

class Solution {
public:
    string findReplaceString(string s, vector<int>& indices, vector<string>& sources, vector<string>& targets) {
        int k = indices.size();
        vector<tuple<int, string*, string*>> zip;
        for(int i=0; i<k; ++i){
            zip.push_back({indices[i], &sources[i], &targets[i]});
        }
        sort(zip.begin(), zip.end());
        string pre = s.substr(0, get<0>(zip.front()));
        vector<string> ss;

        for(int i=0; i<k; ++i){
            ss.push_back(s.substr(get<0>(zip[i]), (i+1<k? get<0>(zip[i+1]) - get<0>(zip[i]): string::npos)));
        }

        for(int i=0; i<k; ++i){
            auto& tmp = zip[i];
            auto &source = *get<1>(tmp), &target = *get<2>(tmp);
            if(ss[i].substr(0, source.size()) == source){
                pre.append(target);
                pre.append(ss[i].substr(source.size()));
            }
            else pre.append(ss[i]);
        }
        return pre;
    }
};
// @lc code=end

File: 833.字符串中的查找与替换.py
#
# @lc app=leetcode.cn id=833 lang=python3
#
# [833] 字符串中的查找与替换
#

# @lc code=start
from typing import List

class Solution:
    def findReplaceString(self, s: str, indices: List[int], sources: List[str], targets: List[str]) -> str:
        indices, sources, targets = zip(*sorted(zip(indices, sources, targets)))
        ss: List[str] = []
        pre = s[: indices[0]]
        for i in range(len(indices)):
            ss.append(s[indices[i]: (indices[i+1] if i+1 < len(indices) else None)])
        ret: List[str] = []
        for subs, source, target in zip(ss, sources, targets):
            if subs.startswith(source):
                ret.append(target + subs[len(source):])
            else:
                ret.append(subs)
        return pre + ''.join(ret)

# @lc code=end

print(Solution().findReplaceString("vmokgggqzp", [3,5,1], ["kg","ggq","mo"], ["s","so","bfr"]))